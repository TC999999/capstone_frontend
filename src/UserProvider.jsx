import UserContext from "./UserContext.js";
import { useState, useEffect } from "react";
import marketAPI from "../api.js";

const UserProvider = ({ children }) => {
  const [user, setUser] = useState(null);

  //sets context for user for entire app
  const currentUser = (user) => {
    setUser(user);
  };

  // //On initial render, get user token from local storage, gets the current user from the api, and set as context
  useEffect(() => {
    async function getToken() {
      let token = localStorage.getItem("market-token");
      if (token) {
        let userInfo = await marketAPI.getCurrentUser();
        currentUser(userInfo);
      }
    }
    getToken();
  }, []);

  //These functions and user state will be used in context
  //Set the token in local storage, retrives the current user from the api, and sets the user as context on login
  const logIn = async (username, token) => {
    localStorage.setItem("market-token", token);
    let userInfo = await marketAPI.getUserInfo(username);
    currentUser(userInfo);
  };

  //remove token and user from local storage and set user context to null on logout
  const logOut = () => {
    localStorage.removeItem("market-token");
    currentUser(null);
  };

  //update current user in local storage and context
  const updateUser = async (username) => {
    let userInfo = await marketAPI.getUserInfo(username);
    currentUser(userInfo);
  };

  return (
    <UserContext.Provider
      value={{ user, currentUser, logIn, logOut, updateUser }}
    >
      {children}
    </UserContext.Provider>
  );
};

export default UserProvider;
