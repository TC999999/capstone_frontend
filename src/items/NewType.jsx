import { useState, useEffect, useContext } from "react";
import UserContext from "../UserContext.js";
import { useNavigate } from "react-router-dom";
import marketAPI from "../../api.js";
import "../styles/NewType.css";

const NewType = () => {
  const initialState = {
    typeName: "",
  };
  const [isLoading, setIsLoading] = useState(false);
  const [formData, setFormData] = useState(initialState);
  const [types, setTypes] = useState([]);
  const { user } = useContext(UserContext);
  const navigate = useNavigate();
  const [err, setErr] = useState(false);
  const [message, setMessage] = useState("");

  useEffect(() => {
    const getTypes = async () => {
      const typeRes = await marketAPI.getAllTypes();
      setTypes(typeRes);
    };

    setIsLoading(true);
    if (user && user.isAdmin) {
      getTypes();
      setIsLoading(false);
    } else {
      setIsLoading(false);
    }
  }, [user]);

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((data) => ({ ...data, [name]: value }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    setIsLoading(true);
    try {
      let { typeName } = formData;

      await marketAPI.addNewItemType({ typeName });

      setFormData(initialState);
      navigate(`/users/${user.username}`);
    } catch (err) {
      setErr(true);
      setMessage(err);
      setIsLoading(false);
    }
  };

  if (isLoading) {
    return <p className="loading-message">Loading...</p>;
  }

  if (!user) {
    return <h1>Please Log In First!</h1>;
  }

  if (!user.isAdmin) {
    return <h1>Only admins can see this page!</h1>;
  }

  return (
    <div className="new-type-page-div">
      <div className="new-type-info-div">
        <div className="current-types-div">
          <h3>Current Types</h3>
          <ul>
            {types.map((type) => {
              return <li key={`type-${type.id}`}>{type.name}</li>;
            })}
          </ul>
        </div>
        <div className="add-new-type-form">
          <h3>Add New Type Here</h3>
          <form onSubmit={handleSubmit}>
            <div className="new-type-form-div new-type-name-div">
              <label htmlFor="typeName"></label>

              <input
                type="text"
                id="typeName"
                name="typeName"
                placeholder="type name of new type here"
                value={formData.typeName}
                onChange={handleChange}
                required
              />
            </div>
            <div className="new-type-form-div">
              <button
                className="add-item-type-form-button"
                onSubmit={handleSubmit}
              >
                Add Type
              </button>
            </div>
          </form>
          {err && <p className="err-msg">{message}</p>}
        </div>
      </div>
      <button className="back-button" onClick={() => navigate(-1)}>
        â¬… Back
      </button>
    </div>
  );
};

export default NewType;
