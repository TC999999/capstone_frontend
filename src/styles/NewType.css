.new-type-info-div {
  display: flex;
  justify-content: space-evenly;
}

.new-type-info-div h3 {
  font-family: Impact, Haettenschweiler, "Arial Narrow Bold", sans-serif;
}

.current-types-div {
  border: solid 2px crimson;
  width: 200px;
  margin: 5px;
  padding: 5px;
  border-radius: 10px;
  box-shadow: 5px 5px 5px black;
  background-color: rgb(253, 120, 120);
}

.current-types-div li {
  font-family: Arial, Helvetica, sans-serif;
  margin: 5px;
}

.add-new-type-form {
  border: solid 2px green;
  border-radius: 10px;
  width: 300px;
  padding: 5px;
  height: 200px;
  margin: 10px;
  box-shadow: 5px 5px 5px black;
  background-color: lightgreen;
  position: sticky;
}

.new-type-name-div input[type="text"] {
  font-size: 15px;
  width: 90%;
}

@media (min-width: 800px) {
  .new-type-name-div input[type="text"] {
    font-size: 18px;
  }
}

.add-item-type-form-button {
  background-color: rgb(36, 209, 243);
  font-weight: bold;
  padding: 10px;
  border: solid 1px;
  box-shadow: 2px 2px 2px black;
  border-radius: 10px;
  margin: 5px;
}

.add-item-type-form-button:hover {
  background-color: blue;
  color: white;
}

.add-item-type-form-button:active {
  background-color: rgb(36, 209, 243);
  color: black;
}
