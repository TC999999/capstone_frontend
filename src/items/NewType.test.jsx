import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import UserProvider from "../UserProvider.jsx";
import NewType from "./NewType.jsx";

it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <UserProvider>
        <NewType />
      </UserProvider>
    </MemoryRouter>
  );
});

it("should match snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <NewType />
      </UserProvider>
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});
