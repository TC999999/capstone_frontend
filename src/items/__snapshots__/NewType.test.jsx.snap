// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <h1>
    Please Log In First!
  </h1>
</DocumentFragment>
`;
